src/basic/fd-util.c

```diff
@@ -615,3 +615,169 @@ int fd_move_above_stdio(int fd) {
+                }
+        }
+
+        if (original_output_fd == STDOUT_FILENO)
+                output_fd = -1; /* Nothing to do, fd is already at the right place */
+        else {
+                if (original_output_fd < 0) {
+
+                        if (original_input_fd < 0)
+                                output_fd = new_input_fd; /* reuse aleady opened /dev/null */
+                        else {
+                                new_output_fd = open("/dev/null", O_RDONLY|O_CLOEXEC);
+                                if (new_output_fd < 0) {
+                                        r = -errno;
+                                        goto finish;
+                                }
+                        }
+                } else
+                        output_fd = original_output_fd;
+
+                if (output_fd < 3) { /* Move out of stdio territory */
+                        int copy;
+
+                        copy = fcntl(output_fd, F_DUPFD_CLOEXEC, 3);
+                        if (copy < 0) {
+                                r = -errno;
+                                goto finish;
+                        }
+
+                        output_fd = new_output_fd = copy;
+                }
+        }
+
+        if (original_error_fd == STDERR_FILENO)
+                error_fd = -1; /* Nothing to do, fd is aleady at the right place */
+        else {
+                if (original_error_fd < 0) {
+
+                        if (original_input_fd < 0)
+                                error_fd = new_input_fd; /* reuse aleady opened /dev/null */
+                        else if (original_output_fd < 0)
+                                error_fd = new_output_fd; /* dito */
+                        else {
+                                new_error_fd = open("/dev/null", O_CLOEXEC|O_RDONLY);
+                                if (new_error_fd < 0) {
+                                        r = -errno;
+                                        goto finish;
+                                }
+                        }
+                } else
+                        error_fd = original_error_fd;
+
+                if (error_fd < 3) { /* Move out of stdio territory */
+                        int copy;
+
+                        copy = fcntl(error_fd, F_DUPFD_CLOEXEC, 3);
+                        if (copy < 0) {
+                                r = -errno;
+                                goto finish;
+                        }
+
+                        error_fd = new_error_fd = copy;
+                }
+        }
+
+        /* At this point we now have the fds to us in input_fd, output_fd and error_fd, and they are all above the
+         * stdio range, so that we have freedom to move them around. If the fds already were at the right places then
+         * the specific fds are -1. Let's now move them to the right places. This is the point of no return. */
+
+        if (input_fd < 0) {
```

Comment:
Here too, why not make this a loop over an array of fds? Repeating the same code three times is error prone.

Reply:
this one is different, indeed.